agent/acs/client/acs_client.go

```diff
@@ -0,0 +1,326 @@
+type requestHandler interface{}
+
+// ClientServer is a combined client and server for the ACS websocket connection
+type ClientServer interface {
+	AddRequestHandler(requestHandler)
+	MakeRequest(input interface{}) error
+	Connect() error
+	Serve() error
+	io.Closer
+}
+
+// default implementation of ClientServer
+type clientServer struct {
+	// url is the full url to ACS, including path, querystring, and so on.
+	url                string
+	region             string
+	credentialProvider credentials.AWSCredentialProvider
+	acceptInvalidCert  bool
+
+	conn websocketConn
+
+	// requestHandlers is a map from message types to handler functions of the
+	// form:
+	//     "FooMessage": func(message *ecsacs.FooMessage)
+	requestHandlers map[string]requestHandler
+}
+
+// receivedMessage is the intermediate message used to unmarshal a message from ACS
+type receivedMessage struct {
+	Type    string          `json:"type"`
+	Message json.RawMessage `json:"message"`
+}
+
+// requestMessage is the intermediate message marshalled to send to ACS.
+type requestMessage struct {
+	Type    string          `json:"type"`
+	Message json.RawMessage `json:"message"`
+}
+
+// New returns a client/server to bidirectionally communicate with ACS
+// The returned struct should have both 'Connect' and 'Serve' called upon it
+// before being used.
+func New(url string, region string, credentialProvider credentials.AWSCredentialProvider, acceptInvalidCert bool) ClientServer {
+	cs := &clientServer{url: url, region: region, credentialProvider: credentialProvider, acceptInvalidCert: acceptInvalidCert}
+	cs.requestHandlers = make(map[string]requestHandler)
+	return cs
+}
+
+// Connect opens a connection to ACS and upgrades it to a websocket. Calls to
+// 'MakeRequest' can be made after calling this, but responss will not be
+// receivable until 'Serve' is also called.
+func (cs *clientServer) Connect() error {
+	parsedAcsURL, err := url.Parse(cs.url)
+	if err != nil {
+		return err
+	}
+
+	signer := authv4.NewHttpSigner(cs.region, acsService, cs.credentialProvider, nil)
+
+	// NewRequest never returns an error if the url parses and we just verified
+	// it did above
+	request, _ := http.NewRequest("GET", cs.url, nil)
+	signer.SignHttpRequest(request)
+
+	// url.Host might not have the port, but tls.Dial needs it
+	dialHost := parsedAcsURL.Host
+	if !strings.Contains(dialHost, ":") {
+		dialHost += ":443"
+	}
+
+	timeoutDialer := &net.Dialer{Timeout: acsConnectTimeout}
+	log.Info("Creating poll dialer", "host", parsedAcsURL.Host)
+	acsConn, err := tls.DialWithDialer(timeoutDialer, "tcp", dialHost, &tls.Config{InsecureSkipVerify: cs.acceptInvalidCert})
+	if err != nil {
+		return err
+	}
+
+	websocketConn, httpResponse, err := websocket.NewClient(acsConn, parsedAcsURL, request.Header, 1024, 1024)
+	if httpResponse != nil {
+		defer httpResponse.Body.Close()
+	}
+	if err != nil {
+		var resp []byte
+		if httpResponse != nil {
+			resp, _ = ioutil.ReadAll(httpResponse.Body)
```

Comment:
I don't think this one can actually blow up badly.
I think the two cases are the following:
1) There's no error here, decodeData unmarshals it and you get a nicely typed error
2) There is an error here, decodeData errors out and you instead get 'errors.New(string(resp))' showing that it failed to ReadAll here.. and you get the websocket error as well.

If you want I can add a comment or handle it, but I don't think it's necessary.